### PR TITLE
JSExpr Rewrite (AST formalization)

### PR DESCRIPTION
**Pull request by twavv**
Continuation of #25 (this branch is inside of JSExpr instead of being inside my own fork).

### CODE REVIEW

**Comment by twavv** [comment]
There are a few things that I think deserve thinking about in a bit more detail before releasing this.

### Extra syntax
I added syntax for objects using braces.
```julia
@js myObj = {
  :foo => "bar",
  :spam => "eggs",
}
# myObj = {["foo"]: "bar",["spam"]: "eggs"}
```

I'm personally slightly in favor of only allowing this syntax and disallowing using `Dict` since that's the only* function we translate from Julia to JS (we don't, for example, translate `println` and I don't think we should either). That would be breaking, though, and the braces syntax might be kind of ugly. The reason I'm for this is that `Dict` is a valid identifier in JS and we've effectively disallowed it (though one can actually do `js"Dict"(...)` to invoke a function named `Dict` in JS), though it's not a huge deal.

\* actually I added a conversion for `:` (range constructor) because that's not a valid JS identifier and it makes writing julia-esque code easier (i.e., `for i in 1:10` becomes a JS loop over 1, ..., 10).

### Macro Behavior
Currently, the macro behavior is to assume that we should just evaluate macros and then interpolate whatever the macro evaluates to. This lets us write macros like `jsshow` which works similarly to Julia's `@show` but logs to the console.
```julia
macro jsshow(ex)
    strrepr = string(ex)
    return :(@js((function (arg)
        console.log($strrepr + " = " + arg); return arg
    end)($(ex))))
end

@js @jsshow foo.bar()
# When evaluates, logs `foo.bar() = ...`
```

The alternative is trying to expand Julia macros and then convert the expanded code into JS, which is not likely to work (since they expanded code will make reference to Julia functions).

This behavior is pretty much equivalent to just interpolation though, so we might just disallow macro expansion and require it to be interpolated which would let us do something later with macro expansion (it's currently possible to override specific macro names, which is how `@var` and friends work (see `crawl_macrocall`).

**tl;dr:** Should we disallow macro calls unless they're explicitly defined for JS (like `@var`, `@new`, ...)?

### Object Iteration
In this version, `for k in v` in Julia is translated to `for k of v` in JS. This works better for arrays since `for x in ["foo", "bar"]` results in `x` taking on the values `1` and `2` since `in` iterates over object keys. However, it doesn't work on objects (`for k of {}` is an error since objects aren't iterable). It's still possible to do `for k of Object.keys({})` which is more idiomatic JS anyway.

This is technically breaking.

**Comment by piever** [comment]
It's a lot of changes, I'll try to give some feedback on the summary (haven't had time to go through the diff yet).

## Extra syntax

> I added syntax for objects using braces.
> 
> ```julia
> @js myObj = {
>   :foo => "bar",
>   :spam => "eggs",
> }
> # myObj = {["foo"]: "bar",["spam"]: "eggs"}
> ```

This syntax really doesn't look ideal. Could we just use named tuples for this, i.e.

```julia
@js myObj = (foo = "bar", spam = "eggs")
# myObj = {["foo"]: "bar",["spam"]: "eggs"}
 ```
 
This already works (on JSExpr release) doing `@js $a` with `a` a named tuple, but fails with `@js (a=1, b=2)`, maybe we could just fix that?

> I'm personally slightly in favor of only allowing this syntax and disallowing using `Dict` since that's the only* function we translate from Julia to JS

Is the proposal to disallow `@js Dict(:a => 1)` but still allow
```julia
a = Dict(:a => 1)
@js $a
```
In this case it would probably not be so bad if we allow named tuples (as mentioned above).

## Macro Behavior

> This behavior is pretty much equivalent to just interpolation though, so we might just disallow macro expansion and require it to be interpolated which would let us do something later with macro expansion (it's currently possible to override specific macro names, which is how `@var` and friends work (see `crawl_macrocall`).

That makes sense, esp. if the objective is to break everything now before v1.0. It would be good to have a plan for the future to see if the breakage is justified (see whether there are practical usecases that are in conflict with the current behavior).

## Object iteration

> In this version, `for k in v` in Julia is translated to `for k of v` in JS.

:100:
Things never iterate their keys by default in julia, so having `for k in v` for arrays and `for k in keys(v)` for named tuples and dicts is probably a good way to code that would work in both languages.

**Comment by twavv** [comment]
Thank you for the detailed response! Fwiw there are a lot of lines but not that much code (a lot is comments, docstrings, and just liberal spacing - most functions are just a single expression over multiple lines).

> This syntax really doesn't look ideal. Could we just use named tuples for this, i.e.

I agree about braces syntax looking weird, but at the same time, objects are such a fundamental part of JS that part of me feels like the syntax should look more distinct than naked tuples (though we can definitely enable named tuples to parse as objects).

**[edit]**
Are there any visually distinct (but also natural?) looking syntaxes we could use?

I will say, using `Dict` is not the **worst** thing in the entire world but I have reservations about only translating a certain few functions, since otherwise, we make zero effort to translate _semantics_ between Julia and JS, only syntax, and I think attempting to do so will be a slippery slope where it's hard to predict what the generated JS code will look like/do. Maybe a `@Dict` macro?
**[/edit]**

> That makes sense, esp. if the objective is to break everything now before v1.0. It would be good to have a plan for the future to see if the breakage is justified (see whether there are practical usecases that are in conflict with the current behavior).

I'm not sure what you're saying. Do you think we should keep macros as I have them implemented or just disable their parsing (excluding `@new`, etc)?

**[edit]**
Currently, the macro parsing doesn't actually evaluate things like `@new` and `@const` (in fact, `@const` is not even a real macro and it's not exported), so it would be inconsistent for some macros to be evaluated and some to be treated as JS keywords, so **I think we should just not attempt to evaluate inline macros** and only allow macros that have an accompanying `crawl_macrocall` method define.
**[/edit]**

> Things never iterate their keys by default in julia, so having for k in v for arrays and for k in keys(v) for named tuples and dicts is probably a good way to code that would work in both languages.

Just to be clear, with this, one would need to use Object.keys, not just keys. I'm kind of conflicted about that (using keys is very similar to Julia and there's less having to switch between Julia and JS modes, but having just a limited subset of functions translated to JS is not a great idea imo). This is similar to my reasoning for not translating Dict (though that one I might renege on).

**Comment by twavv** [comment]
I added namedtuple support (it translates to object syntax in JS).

I don't think this is a complete replacement for a dedicated object syntax both for the reason I described above (I think the object syntax should be visually distinct) as well as the fact that named tuple don't support variable keys.

**i.e.,** this is not possible using namedtuples:
```julia
mykey = "foobar"
Dict(mykey => "myvalue")
# equivalent to Dict("foobar" => "myvalue")
```

This is possible using the braces syntax since it's meant to mirror how `Dict` behaves in Julia.
```julia
@js { mykey => "myvalue" }
# == js"""{ [mykey]: "myvalue" }"""
```

**Comment by piever** [comment]
I see the concern. The other tricky bit is the empty struct, which can't really be expressed easily with `NamedTuple`.

My gripe with `{a => b}` is that it's neither JSON nor the curly bracket syntax used by Query (see [here](https://www.queryverse.org/Query.jl/stable/linqquerycommands/#Example-3), implemented [here](https://github.com/queryverse/Query.jl/blob/master/src/query_translation.jl#L14)). The query curly syntax has the issue you mention (of not allowing variable keys), but has the extra similarity with JSON that `{a, b}` is equivalent to `{a:a, b:b}`, which can be handy in a big structure with many fields as it saves a lot of typing.

Maybe a good compromise would be to accept both "Query style" (i.e. keywords `{a = val}` or `{a, b, c}` for automated naming) and pairs (i.e `{:a => val}`).  Having both keywords and pairs API is consistent with, for example, the `DataFrames` API:

```julia
julia> using DataFrames

julia> DataFrame(a = rand(2))
2×1 DataFrame
│ Row │ a        │
│     │ Float64  │
├─────┼──────────┤
│ 1   │ 0.830304 │
│ 2   │ 0.701768 │

julia> DataFrame(:a => rand(2))
2×1 DataFrame
│ Row │ a         │
│     │ Float64   │
├─────┼───────────┤
│ 1   │ 0.104254  │
│ 2   │ 0.0215448 │
```

**Comment by piever** [comment]
> I will say, using `Dict` is not the worst thing in the entire world but I have reservations about only translating a certain few functions, since otherwise, we make zero effort to translate semantics between Julia and JS, only syntax, and I think attempting to do so will be a slippery slope where it's hard to predict what the generated JS code will look like/do. Maybe a `@Dict `macro?

I now understand the principle. The rule "Translate syntax, but not semantics" naturally implies that if a function name in Julia does not correspond to a function name (or any variable name) in js, this should give an undefined error. Then I imagine we should go for curly bracket syntax and `Object.keys` (these are probably important enough that it is worth it to explain them in the docs).

> Currently, the macro parsing doesn't actually evaluate things like `@new` and `@const` (in fact, `@const` is not even a real macro and it's not exported), so it would be inconsistent for some macros to be evaluated and some to be treated as JS keywords, so I think we should just not attempt to evaluate inline macros and only allow macros that have an accompanying `crawl_macrocall` method define.

I think I agree with this change. I'm wondering, does this allow to put `js"..."` inside of `@js`? Imagine you are writing a big function with `@js` and realize that a subset of it is simpler to write directly in javascript, I guess we should need to make sure that this works.

**Comment by twavv** [comment]
@piever You're incredibly helpful and I'm very grateful.

> My gripe with {a => b} is that it's neither JSON nor the curly bracket syntax used by Query (see here, implemented here). The query curly syntax has the issue you mention (of not allowing variable keys), but has the extra similarity with JSON that {a, b} is equivalent to {a:a, b:b}, which can be handy in a big structure with many fields as it saves a lot of typing.

I'm not tied to any specific syntax. I think using the Query syntax is great (I've just never used it personally which is why I didn't default to that). The `{ a, b, c }` syntax is definitely great (I've thought about it, just hadn't implemented it). It's also the same syntax in JavaScript.

For variable named keys, we could borrow from JS syntax and wrap keys in square brackets if they're not literal names.
```julia
@js { a, b }
# { "a": a, "b": b }
@js { a=a, b }
# {"a": a, "b": b  }

# Variable key interpolation
# This syntax might be (is?) ugly
@js { [a]=a, b }
# { [a]: a, "b": b }

# Pair syntax
@js { :foo => "foo" }
# { "foo": "foo" }
@js { foo => "foo" }
# { [foo]: "foo" }
```

I have conflicting feelings about including the pair syntax, especially if we do include the `[...]`-key syntax, because it does behave differently than the `=` syntax. I think this is obvious when writing "vanilla" Julia code but might be easy to forget deep inside a `@js` macro.

> I think I agree with this change. I'm wondering, does this allow to put js"..." inside of @js? Imagine you are writing a big function with @js and realize that a subset of it is simpler to write directly in javascript, I guess we should need to make sure that this works.

Absolutely! I did have this special-cased at one point. My initial strategy was to expand the macro (using `macroexpand`) and crawl that, but for reasons I outlined above (both difficulty in implementation, as well as that most macros, aren't going to produce remotely valid JS when expanded), I changed that strategy.

We would special case the `@js` and `@js_str` macros using `crawl_macrocall`. I do have a test case for this:
```julia
@test string(@js js"\$"("#my-id")) == raw"""$("#my-id")"""
```

**Comment by piever** [comment]
> I'm not tied to any specific syntax. I think using the Query syntax is great (I've just never used it personally which is why I didn't default to that). The `{ a, b, c }` syntax is definitely great (I've thought about it, just hadn't implemented it). It's also the same syntax in JavaScript.
> 
> For variable named keys, we could borrow from JS syntax and wrap keys in square brackets if they're not literal names.
>
> I have conflicting feelings about including the pair syntax, especially if we do include the [...]-key syntax, because it does behave differently than the = syntax. I think this is obvious when writing "vanilla" Julia code but might be easy to forget deep inside a `@js` macro.

Let me try to least possible alternatives (I think we should choose only one option).

1. not support the syntax and recommend `a = {}; a[key] = val;`

1. use `=` for "hard-coded" name and `=>` for variable name

1. use the JSON `{[key] = val}` syntax

1. go for an infix operator, I personally like `{$key = val}`, as we are somehow interpolating the variable

Option 1 may be the most conservative option as a syntax can be added after 1.0 and we don't have to worry about this now.

Option 2 will be intuitive for users coming from DataFrames that uses the same distinction, but for example [PGFPlotsX](https://kristofferc.github.io/PGFPlotsX.jl/stable/man/options/#The-@pgf-macro-1) uses the `{key = val}` syntax and allows `=>` or `=` indiscriminately, so I agree with you it may be confusing if we change the meaning from one to the other.

I am a bit skeptical on the JSON syntax `{[key] = val}` (option 3) as it is a bit of an odd mixture, I'd imagine we should either support JSON fully and go `{[key] : val}` or find a Julian  syntax for variable key name.

Option 4 may be my current favorite as long as it doesn't create conflicts or confusion with how we interpolate *julia* variables in `@js`.

**Comment by twavv** [comment]
> 1. not support the syntax and recommend a = {}; a[key] = val;

I assume you mean only allow variable-keys using that syntax. I think this might work for now because it's not a super common thing to do anyway. We should mandate that symbols appear on the left-hand side of any pair expressions.

```julia
@js @const myObj = {
  a,
  b="foo",
  :c => "bar",
}
# { "a": a, "b": "foo", "c": "bar" }

# Syntax error: LHS of pair must be string literal or symbol
@js { d => "c" }

# Instead, one must do
@js myObj[d] = "c"
```

> 2. use = for "hard-coded" name and => for variable name

I agree with what you said. This also has semantics that's very in line with how `Dict(...)` works. It's essentially identical to **1.** except that **1.** would error if you did `{ foo => foo }` and **2.** would use the _value_ of `foo` as the key.
```julia
@js { foo }
# { "foo": foo }
@js { foo => "foo" }
# { [foo]: foo }
@js { foo="foo" }
# { "foo": "foo" }
@js { :foo => "foo" }
# { "foo": "foo" }
```

> 3. use the JSON {[key] = val} syntax

We actually could use `@js { foo: bar, [spam]: eggs }` syntax. It gets interpreted as a call to the `:` function (_i.e._, range constructor).

> 4. go for an infix operator, I personally like {$key = val}, as we are somehow interpolating the variable

This is my least favorite (at least with the `$` operator) because it's then impossible to interpolate variables from Julia into the key expression.

### Conclusion
I think I'm between one and two and could be persuaded for three.

**Comment by twavv** [comment]
I think we should just charge blindly forward with option one. It's the least restrictive in terms of what we can do going forward, has the fewest "footguns"*, and I don't want to bikeshed this forever 🙃.

\* At least in Julia, if you accidentally do `Dict(foo => "...")` instead of `Dict(:foo => "...")`, you get an error right then an there, whereas here you wouldn't get an error until you ran the JS code client side.

**Comment by piever** [comment]
> I think we should just charge blindly forward with option one. It's the least restrictive in terms of what we can do going forward, has the fewest "footguns"*, and I don't want to bikeshed this forever :upside_down_face: .

Agreed! Then let's allow "keyword" or an explicit `Pair{Symbol, <:Any}` for now (option 1) and, if we need to, we can always be more permissive (we can remove the error for "not an explicit symbol" at any time, whereas making it error would be breaking).

File: test/objects.jl [hunk 1]
```diff
@@ -0,0 +1,59 @@
+using Test
+using JSExpr
+
+@testset "objects" begin
```
Comment (twavv) [inline test/objects.jl:4]: @piever Does this look good?

Comment (piever) [reply test/objects.jl:4]: Yes!
